<span style="color:green">**Prediction of likes and comments based on features in images**</span><br>
The dataset has the features for each of the picture which was scraped from python<br>
Please refer to:<br> 
1. Instagram scraper Jupyter notebook <br>
2. Feature extraction code using MATLAB<br>
3. Alexnet results using MATLAB

In [7]:
import pandas as pd
import numpy as np

In [8]:
df=pd.read_excel(r'C:\Users\bhavv\OneDrive\Desktop\MMA2\SEMESTER-3\Advanced marketing analytics\final tags.xlsx')

In [9]:
df.columns

Index(['name_of_file', 'location', 'publication_time', 'img_tag', 'cluster',
       'likes_count', 'comments_count', 'Location', 'Brightness_comp1',
       'Brightness_comp2', 'Brightness_comp3', 'Brightness_comp4',
       'Brightness_comp5', 'Color_comp', 'lbp_comp1', 'lbp_comp2', 'lbp_comp3',
       'lbp_comp4', 'lbp_comp5', 'lbp_comp6', 'lbp_comp7', 'lbp_comp8',
       'lbp_comp9', 'lbp_comp10', 'lbp_comp11', 'lbp_comp12', 'lbp_comp13',
       'lbp_comp14', 'lbp_comp15', 'lbp_comp16', 'lbp_comp17', 'lbp_comp18',
       'lbp_comp19', 'lbp_comp20', 'lbp_comp21', 'lbp_comp22', 'lbp_comp23',
       'lbp_comp24', 'lbp_comp25', 'lbp_comp26', 'lbp_comp27', 'lbp_comp28',
       'lbp_comp29', 'lbp_comp30', 'lbp_comp31', 'lbp_comp32', 'lbp_comp33',
       'lbp_comp34', 'lbp_comp35', 'lbp_comp36', 'lbp_comp37', 'lbp_comp38',
       'lbp_comp39', 'lbp_comp40', 'lbp_comp41', 'lbp_comp42', 'lbp_comp43',
       'lbp_comp44', 'lbp_comp45', 'lbp_comp46', 'lbp_comp47', 'lbp_comp48',
       'lbp_comp4

#### Dummify the columns

In [10]:
df=pd.get_dummies(df, columns=['cluster'])

#### Remove nulls

In [20]:
null_columns=df.columns[df.isnull().any()]
df[null_columns].isnull().sum()
df.dropna()

name_of_file    location  \
0    51588789_121604065566862_1242365407126108215_n...  Café OSMO   
1    52859537_641288516295374_2945571047810537898_n...  Café OSMO   
2    53080305_664178464038225_2335123196286281384_n...  Café OSMO   
3    53117361_557998974687563_4872307461780265087_n...  Café OSMO   
4    53250692_399575454192866_213624654316178354_n.jpg  Café OSMO   
..                                                 ...         ...   
283  45623604_263283504536714_631742918939646977_n.jpg  Café OSMO   
284  46120240_332666087518226_6000161391498724686_n...  Café OSMO   
285  46181911_2142728609122969_2012767236307127743_...  Café OSMO   
286  49384987_394681944612574_8004080077372822776_n...  Café OSMO   
287  50230788_291558224863374_3435065724297276701_n...  Café OSMO   

     publication_time  img_tag  likes_count  comments_count      Location  \
0          1549901866        1          127               1    restaurant   
1          1552080122        2           52               1         modem   
2          1554233729        3          126               5         crate   
3          1552083770        4           62               1   book jacket   
4          1552083770        5           62               1   whiskey jug   
..                ...      ...          ...             ...           ...   
283        1543346201      301           89               0  wooden spoon   
284        1542761414      302           72               1      strainer   
285        1544646144      303           72               4    lumbermill   
286        1546807505      304           91               2     shoe shop   
287        1549477679      305           89               5      espresso   

     Brightness_comp1  Brightness_comp2  Brightness_comp3  ...  numhog_comp3  \
0                9409             21252             34912  ...          1591   
1                 218             51617             69518  ...           757   
2               14975            119600             34961  ...          1987   
3               34166             42851             19343  ...          2571   
4               21231             32342             73054  ...          1644   
..                ...               ...               ...  ...           ...   
283             10977             89439             24060  ...          2979   
284             27043             67980             93153  ...          1899   
285             18344            122850             48146  ...           450   
286             23482            105630             52519  ...          2451   
287              9982             30974             31697  ...          3964   

     numhog_comp4  numhog_comp5  numhog_comp6  num_lines  cluster_1  \
0               6             1             3         14          1   
1               0             0             2         10          1   
2              17             0             2          8          1   
3             287            15             3          3          0   
4              25             1             3          7          0   
..            ...           ...           ...        ...        ...   
283            62             1             3          8          0   
284             2             0             3         10          0   
285             0             0             2         15          0   
286            10             0             3         11          0   
287           128            12             3          8          0   

     cluster_2  cluster_3  cluster_4  cluster_5  
0            0          0          0          0  
1            0          0          0          0  
2            0          0          0          0  
3            0          0          1          0  
4            0          1          0          0  
..         ...        ...        ...        ...  
283          0          0          0          1  
284          0          0          0          1  
285         

#### Dummify the prominent column from Alexnet

In [115]:
X_dum = pd.get_dummies(df['prominent'], columns = ['prominent'])

In [116]:
X_dum.head(10)

American lobster  Cardigan  Chihuahua  Crock Pot  Dungeness crab  \
0                 0         0          0          0               0   
1                 0         0          0          0               0   
2                 0         0          0          0               0   
3                 0         0          0          0               0   
4                 0         0          0          0               0   
5                 0         0          0          0               0   
6                 0         0          0          0               0   
7                 0         0          0          0               0   
8                 0         0          0          0               0   
9                 0         0          0          0               0   

   Dutch oven  abaya  academic gown  acorn  acorn squash  ...  vault  \
0           0      0              0      0             0  ...      0   
1           0      0              0      0             0  ...      0   
2           0      0              0      0             0  ...      0   
3           0      0              0      0             0  ...      0   
4           0      0              0      0             0  ...      0   
5           0      0              0      0             0  ...      0   
6           0      0              0      0             0  ...      0   
7           0      0              0      0             0  ...      0   
8           0      0              0      0             0  ...      0   
9           0      0              0      0             0  ...      0   

   vending machine  volleyball  walking stick  wall clock  whiskey jug  wok  \
0                0           0              0           0            0    0   
1                0           0              0           0            0    0   
2                0           0              0           0            0    0   
3                0           0              0           0            0    0   
4                0           0              0           0            1    0   
5                0           0              0           0            0    0   
6                0           0              0           0            0    0   
7                0           0              0           0            0    0   
8                0           0              1           0            0    0   
9                0           0              0           0            0    0   

   wooden spoon  worm fence  zucchini  
0             0           0         0  
1             0           0         0  
2             0           0         0  
3             0           0         0  
4             0           0         0  
5             0           0         0  
6             0           0         0  
7             0           0         0  
8             0           0         0  
9             0           0         0  

[10 rows x 168 columns]

In [22]:
from sklearn.feature_selection import RFE
from sklearn.ensemble import RandomForestRegressor

In [23]:
columns=['publication_time','comments_count', 'Brightness_comp1','Brightness_comp2', 'Brightness_comp3', 'Brightness_comp4','Brightness_comp5', 'Color_comp', 'lbp_comp1', 'lbp_comp2', 'lbp_comp3','lbp_comp4', 'lbp_comp5', 'lbp_comp6', 'lbp_comp7', 'lbp_comp8','lbp_comp9', 'lbp_comp10', 'lbp_comp11', 'lbp_comp12', 'lbp_comp13','lbp_comp14', 'lbp_comp15', 'lbp_comp16', 'lbp_comp17', 'lbp_comp18','lbp_comp19', 'lbp_comp20', 'lbp_comp21', 'lbp_comp22', 'lbp_comp23','lbp_comp24', 'lbp_comp25', 'lbp_comp26', 'lbp_comp27', 'lbp_comp28','lbp_comp29', 'lbp_comp30', 'lbp_comp31', 'lbp_comp32', 'lbp_comp33','lbp_comp34', 'lbp_comp35', 'lbp_comp36', 'lbp_comp37', 'lbp_comp38','lbp_comp39', 'lbp_comp40', 'lbp_comp41', 'lbp_comp42', 'lbp_comp43','lbp_comp44', 'lbp_comp45', 'lbp_comp46', 'lbp_comp47', 'lbp_comp48','lbp_comp49', 'lbp_comp50', 'lbp_comp51', 'lbp_comp52', 'lbp_comp53','lbp_comp54', 'lbp_comp55', 'lbp_comp56', 'lbp_comp57', 'lbp_comp58','lbp_comp59', 'num_corner1', 'numhog_comp1', 'numhog_comp2','numhog_comp3', 'numhog_comp4', 'numhog_comp5', 'numhog_comp6','num_lines']

In [24]:
y=df['likes_count']
X=df.reindex(columns=columns)

##### Feature selection

In [28]:
from sklearn.ensemble import RandomForestRegressor
randomforest = RandomForestRegressor(random_state=0)

model = randomforest.fit(X,y)

from sklearn.feature_selection import SelectFromModel
sfm = SelectFromModel(model, threshold=0.05)
sfm.fit(X,y)

for feature_list_index in sfm.get_support(indices=True):
    print(X.columns[feature_list_index])

feature_df = pd.DataFrame(list(zip(X.columns,model.feature_importances_)), columns = ['predictor','Gini coefficient'])
feature_df.to_csv('RandomForest_FeatureImp.csv')

C:\Users\bhavv\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


publication_time


#### Subset the features which are prominent

In [38]:
X_random_forest=df[['num_corner1','lbp_comp59','numhog_comp6','lbp_comp30','numhog_comp5','lbp_comp53','lbp_comp1','lbp_comp11','lbp_comp27','numhog_comp1','lbp_comp2','lbp_comp29','numhog_comp2','lbp_comp20','lbp_comp24','lbp_comp17','lbp_comp58','lbp_comp49','lbp_comp9','lbp_comp57','lbp_comp55','lbp_comp15','lbp_comp10','lbp_comp51','Brightness_comp2','lbp_comp26','lbp_comp5','lbp_comp16','lbp_comp43','lbp_comp48','lbp_comp12','lbp_comp6','lbp_comp13','lbp_comp54','num_lines','lbp_comp8','lbp_comp33','lbp_comp46','lbp_comp47','lbp_comp50','lbp_comp25','lbp_comp28','lbp_comp22','lbp_comp36','lbp_comp4','lbp_comp41','lbp_comp14','lbp_comp45','lbp_comp39','lbp_comp21','lbp_comp56','lbp_comp32','lbp_comp42','lbp_comp34','Brightness_comp5','numhog_comp3','lbp_comp18','Brightness_comp3','lbp_comp38','lbp_comp31','lbp_comp44','comments_count','lbp_comp19','lbp_comp35','lbp_comp23','lbp_comp52','Brightness_comp1','lbp_comp37','lbp_comp7','lbp_comp40','lbp_comp3','numhog_comp4','Brightness_comp4','Color_comp','publication_time']]

#### Correlation matrix

In [54]:
corr = X_random_forest.corr()
corr.style.background_gradient(cmap='coolwarm')

<div class="alert alert-block alert-info">
<b>Regression model for #Likes as Y-variable</b>
</div>

#### Subset features

In [147]:
X_selected1=df[['num_corner1','numhog_comp6','numhog_comp5','numhog_comp1','numhog_comp2','lbp_comp58','Brightness_comp2','lbp_comp6','lbp_comp54','num_lines','lbp_comp33','lbp_comp50','lbp_comp25','lbp_comp28','lbp_comp22','lbp_comp36','lbp_comp4','lbp_comp14','lbp_comp39','lbp_comp21','lbp_comp56','lbp_comp32','lbp_comp34','Brightness_comp5','numhog_comp3','lbp_comp18','Brightness_comp3','lbp_comp38','lbp_comp31','lbp_comp44','comments_count','lbp_comp19','lbp_comp35','lbp_comp23','lbp_comp52','Brightness_comp1','lbp_comp37','lbp_comp7','lbp_comp40','lbp_comp3','numhog_comp4','Brightness_comp4','Color_comp','publication_time']]

#### Adding dummified columns of 'prominent' from Alexnet results gives a higher MSE, hence we have removed it on the models below

In [125]:
X_selected = pd.concat([X_selected1, X_dum], axis=1)

In [126]:
y=df['likes_count']

## Model 1: Random Forest

In [138]:
# Load libraries

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import pandas

# Separate the data
from sklearn.model_selection import train_test_split
X_regtrain_train, X_regtrain_test, y_regtrain_train, y_regtrain_test = train_test_split(X_selected1, y, test_size=0.3, random_state=5)

#### Play with hyperparameters

#### 1. Max Features

In [140]:
#Play with max_features

max_mse = 999999999999999999999999999999
best_i = 0
for i in range(2,20):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, max_features = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [141]:
best_i

17

#### best_i value is 17

#### 2. Max Depth

In [142]:
#Play with max_depth
max_mse = 999999999999999999999999999999
best_i = 0
for i in range(2,50):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, max_depth = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [143]:
best_i

13

#### best_i value is 13

#### 3. min_samples_split

In [144]:
#Play with min_samples_split
max_mse = 9999999999999
best_i = 0
for i in range(2,50):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, min_samples_split = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [145]:
best_i

2

#### best_i value is 2

#### 4. min_samples_leaf

In [134]:
#Play with min_samples_leaf
max_mse = 9999999999999
best_i = 0
for i in range(2,50):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, min_samples_leaf = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [135]:
best_i

2

#### best_i value is 2

#### 5. bootstrap

In [78]:
#Play with bootstrap
max_mse = 9999999999999
best_i = 0
for i in range(0,2): 
    rf = RandomForestRegressor(random_state=0, n_estimators=100, bootstrap = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [79]:
best_i

1

#### bootstrap value = 1

In [139]:
# With all hyperparameters
reg_model_RF = RandomForestRegressor(random_state = 0, max_features = 17, max_depth = 13, min_samples_split = 2, min_samples_leaf = 2, bootstrap = 1, n_estimators = 100)
reg_model_fit= reg_model_RF.fit(X_regtrain_train, y_regtrain_train)
y_regtrain_test_pred= reg_model_fit.predict(X_regtrain_test)

from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
print(mse)

437.41924967233143


#### MSE = 437.41
#### RMSE = 20.9

## Model 2: KNN

In [86]:
# Load libraries
from sklearn.neighbors import KNeighborsRegressor
import pandas

from sklearn.preprocessing import StandardScaler
standardizer= StandardScaler()
X_regtrain_std = standardizer.fit_transform(X_selected)

# Separate the data
from sklearn.model_selection import train_test_split
X_regtrain_train, X_regtrain_test, y_regtrain_train, y_regtrain_test = train_test_split(X_regtrain_std, y, test_size=0.33, random_state=3)

# Run K-NN
max_mse = 0
max_i = 0

from sklearn.metrics import mean_squared_error
#i is number of neighbours
for i in range(2,5):
    knn= KNeighborsRegressor(n_neighbors=i)
    model = knn.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred = knn.predict(X_regtrain_test)
    mse=mean_squared_error(y_regtrain_test, y_regtrain_test_pred)

    if mse > max_mse:
        max_mse = mse
        max_i = i
        
print(max_mse)
print(max_i)

1008.0381944444445
3


#### MSE = 1008.03
#### RMSE = 31.75

## Model 3: Linear Regression

In [90]:
# Load libraries
from sklearn.linear_model import LinearRegression
import pandas

# Import data
# Separate the data
from sklearn.model_selection import train_test_split
X_regtrain_train, X_regtrain_test, y_regtrain_train, y_regtrain_test = train_test_split(X_regtrain_std, y, test_size =0.3, random_state=3)

# Run linear regression
lm= LinearRegression()
model1 = lm.fit(X_regtrain_train, y_regtrain_train)

# Using the model to predict the results based on the test dataset
y_regtrain_test_pred = lm.predict(X_regtrain_test)

# Calculate the mean squared error of the prediction
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)

print(mse)

797.7407924541571


#### MSE = 797.74
#### RMSE = 28.23

<span style="color:red">Random forest is the best model with a RMSE of 20.9</span>

<div class="alert alert-block alert-info">
<b>Regression model for #Comments as Y-variable</b>
</div>

In [110]:
y=df['comments_count']

## Model 1: Random Forest

In [111]:
# Load libraries

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
import pandas

# Separate the data
from sklearn.model_selection import train_test_split
X_regtrain_train, X_regtrain_test, y_regtrain_train, y_regtrain_test = train_test_split(X_selected, y, test_size=0.3, random_state=5)

#### Play with hyperparameters

#### 1. Max Features

In [98]:
#Play with max_features

max_mse = 999999999999999999999999999999
best_i = 0
for i in range(2,20):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, max_features = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [99]:
best_i

19

#### best_i value is 19

#### 2. Max Depth

In [100]:
#Play with max_depth
max_mse = 999999999999999999999999999999
best_i = 0
for i in range(2,50): #152
    rf = RandomForestRegressor(random_state=0, n_estimators=100, max_depth = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [101]:
best_i

3

#### best_i value is 3

#### 3. min_samples_split

In [102]:
#Play with min_samples_split
max_mse = 9999999999999
best_i = 0
for i in range(2,50):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, min_samples_split = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [103]:
best_i

15

#### best_i value is 15

#### 4. min_samples_leaf

In [104]:
#Play with min_samples_leaf
max_mse = 9999999999999
best_i = 0
for i in range(2,50):
    rf = RandomForestRegressor(random_state=0, n_estimators=100, min_samples_leaf = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [105]:
best_i

2

#### best_i value is 2

#### 5. bootstrap

In [106]:
#Play with bootstrap
max_mse = 9999999999999
best_i = 0
for i in range(0,2): 
    rf = RandomForestRegressor(random_state=0, n_estimators=100, bootstrap = i)
    model= rf.fit(X_regtrain_train, y_regtrain_train)
    y_regtrain_test_pred= rf.predict(X_regtrain_test)
    mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
    if mse < max_mse:
        max_mse = mse
        best_i = i

In [107]:
best_i

1

#### bootstrap value = 1

#### Final Model

In [109]:
# With all hyperparameters
reg_model_RF = RandomForestRegressor(random_state = 0, max_features = 19, max_depth = 3, min_samples_split = 15, min_samples_leaf = 2, bootstrap = 1, n_estimators = 100)
reg_model_fit= reg_model_RF.fit(X_regtrain_train, y_regtrain_train)
y_regtrain_test_pred= reg_model_fit.predict(X_regtrain_test)

from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_regtrain_test, y_regtrain_test_pred)
print(mse)

0.4542930072205883


#### On an average, the posts get 2 comments, hence this is not a really good variable to be predicted

## Thank you!